Add types skeleton of coin selection and largestFirst algo implementation

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#92 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added coin selection proposal for coin selection and fee calculation


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
[92] Add types skeleton and functions to support in the next steps

[92] move to Cardano.Wallet.*

[92] Comply with review

[92] Use Quantity

[92] Remove random and remove other undefined

Anviking commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+-- | A linear equation on the transaction size. Represents the @\s -> a + b*s@
+-- function where @s@ is the transaction size in bytes, @a@ and @b@ are
+-- constant coefficients.
+data TxSizeLinear = TxSizeLinear !Coeff !Coeff
```
_Could_ do this:
```suggestion
data TxSizeLinear = TxSizeLinear !(Quantity "lovelace/byte" Nano) !(Quantity "lovelace" Nano)
```

paweljakubas replied:
no `TxSizeLinear` use any more

Anviking commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+data CoinSelFinalResult = CoinSelFinalResult
+    { csrInputs  :: NonEmpty (TxIn, TxOut)
+    -- ^ Picked inputs
+    , csrOutputs :: NonEmpty TxOut
```
`_outputs` or `outputs` instead? Have we achieved consensus on this yet?

paweljakubas replied:
corrected everywhere

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+data ExpenseRegulation =
+      SenderPaysFee
+      -- ^ The sender pays for the fee. This is the typical case.
+    | ReceiverPaysFee
```
This branch (and whole `ExpenseRegulation` type) can go, we only consider `SenderPaysFee`.

paweljakubas replied:
removed `ExpenseRegulation`

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+    -- ^ UTxO exhausted during input selection
+    -- We record the balance of the UTxO as well as the size of the payment
+    -- we tried to make.
+    -- See also 'CoinSelHardErrCannotCoverFee'
```
We'll probably want to review this ADT, we had quite many cases in the legacy code of types not mapped to the right error and us, remapping after in a wonky way.

paweljakubas replied:
starting with one constructor ADT

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+--                        Fee related                                     --
+----------------------------------------------------------------------------
+
+newtype Fee = Fee { getFee :: Coin }
```
Let's not use `Coin` to model fee. It's actually _wrong_ since fee could in theory be bigger than one `Coin` ...

paweljakubas replied:
used : 
```
newtype Fee = Fee { getFee :: forall a. (Eq a, Ord a) => a }
```

KtorZ replied:
What about `Quantity "lovelace" Natural` ?

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+  :: ByteString
+    -- ^ Maximum size of a transaction
+  -> Word64
+estimateMaxTxInputs _ = undefined
```
Not needed now. We can pretty much simplify this at the moment by considering:

```hs
estimateFee :: Tx -> Word64

-- or

estimateFee :: Int -> Word64 -> Word64 -- a record type to disambiguate parameters would be good
```

which we can `mock` for now.

paweljakubas replied:
done

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+-- function where @s@ is the transaction size in bytes, @a@ and @b@ are
+-- constant coefficients.
+data TxSizeLinear = TxSizeLinear !Coeff !Coeff
+    deriving (Eq, Ord, Show, Generic)
```
Same as above, not needed for the coin selection so-to-speak, this is about fee estimation.

paweljakubas replied:
removed

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+type CoinSelPolicy utxo m a =
+       NonEmpty TxOut
+    -> utxo
+    -> m (Either CoinSelHardErr a)
```
Let's avoid the type-alias here and inline this.

paweljakubas replied:
done

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+random
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> MaximumNumOfInputs
```
`MaximumNumOfInputs` could just be part of `CoinSelectionOptions` :thinking:

paweljakubas replied:
good remark, now `MaximumNumOfInputs` is part of `CoinSelectionOptions`

KtorZ commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,154 @@
+    => CoinSelectionOptions
+    -> MaximumNumOfInputs
+    -> CoinSelPolicy UTxO m CoinSelFinalResult
+largestFirst _ = undefined
```
If we inline the `CoinSelPolicy`, this basically becomes:

```hs
largestFirst 
    :: MonadRandom m
    => CoinSelectionOptions
    -> UTxO 
    -> ExceptT ErrCoinSelection m CoinSelFinalResult
```

paweljakubas replied:
done

akegalj commented on src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -0,0 +1,97 @@
+
+estimateMaxTxInputs
+  :: ByteString
+    -- ^ Maximum size of a transaction
```
size of tx in bytestring - how should one interpret this ?

paweljakubas replied:
you are right, should be `Bytes` (memory unit), like in http://hackage.haskell.org/package/serokell-util-0.10.0/docs/src/Serokell.Data.Memory.Units.html#Byte

paweljakubas replied:
I wonder if we are ok with using `serokell-utils` or better define `Bytes` ourselves?

KtorZ replied:
It's okay to use libraries (from Serokell or anyone) when they are sound and solve an actual problem. Here, `serokell-util` isn't really sound nor useful for most of the types or functions it provides IMO. 

Here, a size could just be a `Quantity "byte" Word32` if you really want to disambiguate the type. Yet, I believe it's not needed to implement largest first :thinking: ?

paweljakubas replied:
if it is not needed in the end I will definitely remove it. Otherwise, will rely on `Quantity "byte" Word32`

paweljakubas committed:
[92] First largestFirst implementation

[92] Compile fix

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,232 @@
+    let txOutputsSorted = NE.toList
+            $ NE.sortBy (flip $ comparing coin) txOutputs
+
+    -- Step 2. (TO-DO or not) we need to check if the transaction outputs are not redeemable
```
what "redeemable" means here?

KtorZ committed:
simplify coin selection implementation

- Less intermediate types
- Rely on built-in function on list instead of custom function for performances (O(n*log(n)) instead of (O(n))
- Correct reporting of errors

paweljakubas committed:
[92] adding coin selection unit tests

paweljakubas committed:
fix largest first errors, add UtxoNotEnoughFragmented and improve on tests

paweljakubas committed:
Adding test for small MaximumNumberOfInputs

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,112 @@
+    when (utxoBalance < moneyRequested)
+        $ throwE $ UtxoExhausted utxoBalance moneyRequested
+
+    -- FIXME ? we need to check if the transaction outputs are not redeemable
```
I wondered yesterday already, but what "redeemable" means in this context?

paweljakubas replied:
there was a check in the legacy code https://github.com/input-output-hk/cardano-wallet-legacy/blob/develop/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs#L292
Basically, checking if the address in transaction output is redeemable

KtorZ replied:
Ah. Okay. I think we don't _care_ here:

1/ We can't generate redeem address anyway
2/ If, through manual selection or whatever process we end up constructing a transaction to a redeem address, the node will reject it. 

I am really not sure what's the rational is here :thinking: ...

paweljakubas replied:
removed unneeded comment

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,112 @@
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
+
+    when (utxoBalance < moneyRequested)
+        $ throwE $ UtxoExhausted utxoBalance moneyRequested
```
I'd actually swap those two checks, such that, in case there's no UTxO available, it still returns a `UTxOExhausted` error (which we could rename to `NotEnoughMoney` basically? Don't you think? The utxo can be exhausted in two ways, as described in a comment, so perhaps, it's better to have this error reflecting this as well).

KtorZ replied:
Also, I think that, those checks where done in the `Nothing` branch previously, such that they will only occur _after_ the coin selection has been tried once. I think the rational was that successful calls to this function are more likely than error calls and thus, rather than doing those checks on every call, we simply do it in case of error and save some CPU cycles along the way.

paweljakubas replied:
swapped the errors, checked it in unit tests, and changed the name of the error

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,112 @@
+                (inp, out):utxo' = utxo
+                target' =
+                    (fromIntegral (getCoin (coin txout)))
+                    - (fromIntegral (getCoin (coin out)))
```
This seems wrong? The new target should be based on the current target and not on the initial one?

paweljakubas replied:
we have `foldM atLeast utxos outputs` and in the `atLeast` we process every output per go. And inside, we are looking which first coin from utxo (previously taken the biggest maxnumberofinputs of them and then taken from the smallest) can cover this target. To my best knowledge, the target within go (I changed inner loop that searches for this to `pickTheFirst`) my stay intact and the above expression does its job.

KtorZ replied:
Hmmm... that isn't correct. The `go` may actually picks more than one UTxO (possibly depleting the whole list if there aren't enough to cover for the target), that's why it's an iterative approach.

The target starts being the txout's value. And then, after every pick, we makes it smaller by the amount of the corresponding UTxO entry we've picked, and iterate, until the target is covered. So, in the last branch, we should have:

```hs
targets' = target - fromIntegral (getCoin (coin out))
```

paweljakubas replied:
ok, finally got it - corrected the algorithm and included boundary and exemplary cases in unit tests to show the implementation is right now

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,280 @@
+        let check =
+                invariant "utxo must cover all transaction outputs"
+                (NE.length txOuts)
+                (\_ -> condCoinsCovering txOuts utxo)
```
You could use a precondition for this using QuickCheck's `(==>)`

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,280 @@
+        it "works as expected for zero-fee and not covered by utxo coin selection"
+            (property propLargestFirstNotCovered)
+        it "works as expected for zero-fee and fully covered by utxo coin selection when maximumNumberOfInputs is small"
+            (property propLargestFirstFullyCoveredSmallMaxInput)
```
I am a puzzled by the properties here :thinking: ? What properties are we actually looking for? I think you're mis-using property testing here. The testing logic is almost as complex as the implementation itself which is a good sign that tests are maybe doing too much. A good indicator is the way you can express the property in prose, what _means_ "works as expected" here, if you're checking for a property, then you should be able to express it.

I'd suggest the following approach:

- Reword these test cases into plain unit tests, where, given some inputs (that you can construct by hand, in a easy-to-read fashion), you obtain some expected output / behavior.

- Add some property-based tests for actual _properties_ of that largest first heuristic, for instance, I can see a few properties that I'd expect on such function:
    - forall (UTxO, NonEmpty TxOut), running largest first twice one the same UTxO yields exactly the same result
    - forall (UTXO, NonEmpty TxOut), for all selected input, there's no bigger input in the UTxO that is not already in the selected inputs. 
    - forall (UTxO, NonEmpty TxOut), there's at least as many selected inputs as there are requested outputs

So in practice, the properties are very short to express, and it's really about performing the coin selection, and then, verifying that the output satisfies some conditions. All properties above assume that there are enough inputs to cover for all of the NonEmpty TxOut, so likely, they'll need preconditions (which could just be `isRight $ runExceptT $ largestFirst ... ==> ...`)

paweljakubas replied:
fair argument. Added unit tests already. Properties will be added in the next commit

paweljakubas replied:
this one is I do not understand  : `forall (UTXO, NonEmpty TxOut), for all selected input, there's no bigger input in the UTxO that is not already in the selected inputs.`

algo goes : we take **maxNumberOfInputs** biggest utxo entries, and then take from the smallest to cover transaction outputs taking from the smallest. It seems, but I may be wrong, that it contradicts ^^^

paweljakubas replied:
I sticked to `CoveringCase` to run it quicker, as correct subset is significantly smaller to whole set of (Utxo, Empty TxOut) - that's my intuition to be verified

KtorZ replied:
> algo goes : we take maxNumberOfInputs biggest utxo entries, 

Yes.

>  and then take from the smallest to cover transaction outputs taking from the smallest

No. We do cover biggest outputs first, and use the biggest inputs first to do so.

paweljakubas replied:
right !!!

paweljakubas replied:
added the property. in the lights of correct implementation it makes sense

KtorZ reviewed (changes requested):
Some good ideas in the test scenarios that would be better expressed as plain hand-written unit tests IMO;

paweljakubas committed:
Adding unit tests and changing to NotEnoughMoney

paweljakubas committed:
Correcting properties

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,113 @@
+
+    -- FIXME ? we need to check if the transaction outputs are not redeemable
+
+    case foldM atLeast (L.reverse $ nLargest utxo, mempty) txOutputsSorted of
```
Why is the list reversed here :thinking: ?

paweljakubas replied:
we start to consume utxo from the smallest cut obtained by maximumNumberInputs (which takes those numbers of biggest utxos entries) . taken from : https://github.com/input-output-hk/cardano-wallet-legacy/blob/develop/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs#L215

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,294 @@
+    -> NonEmpty Word64
+    -- ^ transaction outputs
+    -> Either CoinSelectionError [Word64]
+    -- ^ expecting CoinSelectionError or coins selected
```
Nice. 
What about passing all the arguments above in a record type (instead of having comments here), so that at the call-site, it becomes even more readable ?

paweljakubas replied:
done

paweljakubas committed:
Correct largestFirst, add property and number of tests. Plus code cleaning

just small correction

complying with review

paweljakubas commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,109 @@
+atLeast (utxo0, selection) txout =
+    pickTheFirst (fromIntegral $ getCoin $ coin txout, mempty) utxo0
+    where
+    pickTheFirst
```
this should be called `cover` probably to better reflect intent

paweljakubas commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+            Map.toList (getUTxO utxo) L.\\ utxoEntriesInCoinSelection
+    let getExtremumValue f = f . map (getCoin . coin . snd)
+
+    if L.null utxo' then utxo' `shouldBe` utxo' else
```
is there Expectation that is "True"/expectationSatisfied? why this checking? sometimes utxo is completely depleted after coin selection and minimum from [] does not make sense...maybe secure min with default 0...

KtorZ replied:
Hmmm.. since you're in a `Monad`, you could use `when` or `unless` here (from `Control.Monad`)

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+                 100
+                 [10,10,17]
+                 (17 :| [])
+                 (Right [17])
```
:sweat_smile:  ... the suggestion of introducing a record type earlier was to use the named fields exactly here!

```hs
            coinSelectionUnitTest UnitTestInput
                { maxNumOfInputs = 100
                , utxoInputs = [10,10,17]
                , txOutputs = 10 :| []
                , expectedResult = Right [17]
                }
```

This is where we get most of the readability (at the call-site) :)

paweljakubas replied:
indeed, corrected

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+                 [12,10,17]
+                 (1 :| [])
+                 (Right [17])
+                ))
```
What's the added value for this case compare to the previous one :thinking: ?

paweljakubas replied:
one is smaller than any input coin and the second is the biggest possible to cover by one input - it help me when checking algo impl

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+spec :: Spec
+spec = do
+    describe "Coin selection : LargestFirst algorithm unit tests" $ do
+        it "happy path result in successful selection - one input per output - a" $ do
```
I'd suggest to remove "happy path result in successful selection", and just keep the second part :+1:

paweljakubas replied:
done

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+                 [1,2,10,6,5]
+                 (11 :| [1])
+                 (Left $ MaximumInputsReached 2)
+                ))
```
Good scenarios overall :+1

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+                 Properties and unit test generic scenario
+-------------------------------------------------------------------------------}
+
+data UnitTestInput = UnitTestInput
```
neat-picking but I'd just call that `Fixture`, because that's what it is.

paweljakubas replied:
done

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+    (utxo,txOuts) <- setup
+
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions n)
```
No variable-length indentation please :pray:

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+        Right expectedCoinsSel ->
+            (right (\(CoinSelection inps _  _) -> map (getCoin . coin . snd) inps) result)
+            `shouldBe`
+            (return expectedCoinsSel)
```
What about writing this _the other way around_: transform the result first, and then expect something on it.

```hs
    result <- runExceptT $ do
        CoinSelection inps _ _ <-
            largestFirst (defaultCoinSelectionOptions n) utxo txOuts
        return $ map (getCoin . coin . snd) inps
    result `shouldBe` expected
```

It's usually a good idea to first perform any required transformations on some underlying data-structure before actually running any side-effects.

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+            txUtxoOutsAddr <- generate $ vectorOf (L.length utxoCoins) arbitrary
+            let utxo = UTxO $ Map.fromList $ L.zip txUtxoIns
+                       $ L.zipWith TxOut txUtxoOutsAddr
+                       $ map Coin utxoCoins
```
no variable-length indentation please :pray:

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+                       $ map Coin utxoCoins
+            txOutsAddr <- generate $ vectorOf (L.length txOutsCoins) arbitrary
+            let txOuts = NE.zipWith TxOut (NE.fromList txOutsAddr)
+                         $ NE.map Coin txOutsCoins
```
same here, no variable indentation :pray:

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+        setup :: IO (UTxO, NonEmpty TxOut)
+        setup = do
+            txUtxoIns <- generate $ vectorOf (L.length utxoCoins) arbitrary
+            txUtxoOutsAddr <- generate $ vectorOf (L.length utxoCoins) arbitrary
```
names could just be `ins` and `addrs`, there's not much ambiguity in this context, so shorter-name are more readable :+1:

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+    resultOne <- runExceptT $ largestFirst
+                 (defaultCoinSelectionOptions 100)
+                 utxo
+                 txOuts
```
indentation... :pray:

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+              txOuts
+
+    L.length
+        (concatMap (\ (CoinSelection inps _ _) -> inps) $ rights [result])
```
The usage of `concatMap` + `rights [result]` is a bit weird here and would yield weird failure reports in case of failure. Instead, I'd suggest to perform a pattern match before and leave a clear error message about the outcome. e.g.:

```hs
    let opts = defaultCoinSelectionOptions 100
    nInps <- runExceptT (largestFirst opts utxo txOuts) >>= \case
        Left err -> fail $ "expected success but failed: " <> show err
        Right (CoinSelection inps _ _) -> return $ length inps
    nInps `shouldSatisfy` (>= NE.length txOuts)
```

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+    :: CoveringCase
+    -> Property
+propDeterministic (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
```
We probably don't need the precondition on that property ? I'd expect failures to also be deterministic.

paweljakubas replied:
removed this precondition

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
```
:pray:

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+              txOuts
+    let utxoEntriesInCoinSelection =
+            concatMap (\ (CoinSelection inps _ _) -> inps)
+            $ rights [result]
```
Same remark as before about `concatMap` + `right [..]`

paweljakubas replied:
done

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+            concatMap (\ (CoinSelection inps _ _) -> inps)
+            $ rights [result]
+    let utxo' =
+            Map.toList (getUTxO utxo) L.\\ utxoEntriesInCoinSelection
```
We do have some useful primitive functions in the `Primitive.Types` for UTxO manipulation. This really is:

```hs
-- ins⋪ u
excluding :: UTxO -> Set TxIn ->  UTxO
```

paweljakubas replied:
used `excluding`

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+
+    if L.null utxo' then utxo' `shouldBe` utxo' else
+        (getExtremumValue L.minimum utxoEntriesInCoinSelection)
+        `shouldSatisfy` (>= (getExtremumValue L.maximum utxo'))
```
So, this checks that, the minimal selected value is greater than the maximum value from the remaining utxo. Sounds about right!

paweljakubas replied:
used `unless`

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+                     $ (Map.toList . getUTxO) utxo
+        rotate :: Int -> [a] -> [a]
+        rotate _ [] = []
+        rotate n xs = zipWith const (drop n (cycle xs)) xs
```
I'd suggest a different approach here to avoid having too much logic in the test:

- change the precondition to something like `isRight . runExceptT . largestFirst ...`. In the end, that's really what the pre-condition is about here, checking that we can perform the coin selection! So, let's just perform it and assume that it's right. On small sets of value, it should be fast enough to not be too slow to evaluate with quickcheck.

paweljakubas replied:
really do not know how to use properly `isRight` with `runExceptT` ... but when I use `CoverCase` then it is not need I bet...but drawback is that logic stays

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+
+instance Arbitrary CoveringCase where
+    arbitrary = do
+        n <- choose (1, 2)
```
perhaps slightly more than 2 ? What do you think ?

paweljakubas replied:
now `(1,10)` and works quickly

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,336 @@
+    arbitrary = do
+        n <- choose (1, 2)
+        txOutsNonEmpty <- NE.fromList <$> vectorOf n arbitrary
+        utxo <- arbitrary `suchThat` (condCoinsCovering txOutsNonEmpty)
```
Hmmm... we should have this either as a precondition or in the generator. Having it in both places is redundant. I'd be in favor of generating only valid `CoveringCase`, so checking only here once (previous comment about simplifying the precondition still applies though).

paweljakubas replied:
I like it to have in generator, but when I would know how to put `isRight` in the property precondition I would remove `suchThat` and `condCoinsCovering` code. On the other case, I would need to make sure significant number of covering cases are tried => checkCoverage to be used

KtorZ reviewed (changes requested):
Good tests. I haven't yet re-reviewed the implementation, as I wanted to put the emphasizes on the tests first. 
Some remarks on the form and coding style however.

paweljakubas commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,111 @@
+            throwE $ MaximumInputsReached (fromIntegral n)
+
+
+-- Selecting coins to cover at least the specified value
```
Added this extensive comment how the largestFirst works - hope it will save time anyone trying to understand this algorithm. This along with tests should explain the matter in a comprehensible way

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,111 @@
+--     UTxO input, not participating in the coverage, is taken. We are back at (b)
+--     step as a result
+--
+-- The steps are continued until all transaction are covered.
```
Nice :) 

I'd even add either, before (c) or at the end of (b) that, each output is treated independently wit the heuristic described in (c).

paweljakubas replied:
added

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,111 @@
+--      which coin selection will be tried
+-- (c) the biggest candidate UTxO input is tried first to cover the transaction
+--     output. If the input is not enough, then the next biggest one is added
+--     to check if they can cover the transaction output. This process id continued
```
**id** continued -> is continued

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,331 @@
+        Left err ->
+            result `shouldBe` (Left err)
+        Right expectedCoinsSel ->
+            result `shouldBe` (return expectedCoinsSel)
```
The 5 lines above are strictly equivalent to 

```hs
result `shouldBe`expected
```

The pattern-match is superfluous here.

paweljakubas replied:
indeed!

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,111 @@
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadIO m
```
`IO` isn't needed here, the constraint is _too strong_ and I'd suggest to lower it down to `Monad m`

paweljakubas replied:
right

KtorZ committed:
simplify pre-conditions for properties in CoinSelection

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,109 @@
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. Monad m
```
There was no need for a `MonadIO m =>` constraint here (was actually too strong), so I've replaced it with a `Monad m =>`

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,109 @@
+        $ throwE $ NotEnoughMoney utxoBalance moneyRequested
+
+    when (numberOfUtxoEntries < numberOfTransactionOutputs)
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
```
I'd move these (+ the corresponding local declarations above) into the `Nothing ->` branch, to avoid computing them on every run, but only when the selection failed.

paweljakubas replied:
done

KtorZ commented on src/Cardano/Wallet/CoinSelection/LargestFirst.hs:
```diff
@@ -0,0 +1,109 @@
+                target' =
+                    target - (fromIntegral (getCoin (coin out)))
+            in
+                coverOutput (target', ins ++ [(inp, out)]) utxo'
```
In general, I'd suggest to always prepend elements to lists when using recursion with a list accumulator, and in the terminal case, reverse the list. Because built-in lists in Haskell are linked-list, adding elements to the end of a list is costly, while prepending elements is `O(1)`. Here, it probably doesn't matter much because inputs will remain rather small in practice; but still :no_mouth:

paweljakubas replied:
done, and also updated tests for it

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,288 @@
+    let opts = defaultCoinSelectionOptions 100
+    let resultOne = runIdentity $ runExceptT $ largestFirst opts utxo txOuts
+    let resultTwo = runIdentity $ runExceptT $ largestFirst opts utxo txOuts
+    resultOne === resultTwo
```
changed here the usage of `MonadIO + shouldBe` with just `===` & `runIdentity`; there's no need for full IO :)

paweljakubas replied:
:+1:

KtorZ commented on test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs:
```diff
@@ -0,0 +1,288 @@
+    :: CoveringCase
+    -> Property
+propAtLeast (CoveringCase (utxo, txOuts)) =
+    isRight selection ==> let Right s = selection in prop s
```
So here I've required that selection is Right as a precondition because our property only matters for valid selections.

paweljakubas replied:
beautiful :100:

paweljakubas committed:
Final cleanup and improvements
